pkg/graveler/graveler.go

```diff
@@ -2226,6 +2229,68 @@ func (g *Graveler) dropTokens(ctx context.Context, tokens ...StagingToken) {
 	}
 }
 
+func (g *Graveler) ResetHard(ctx context.Context, repository *RepositoryRecord, branchID BranchID, ref Ref, opts ...SetOptionsFunc) error {
+	isProtectedCommit, err := g.protectedBranchesManager.IsBlocked(ctx, repository, branchID, BranchProtectionBlockedAction_COMMIT)
+	if err != nil {
+		return err
+	}
+	if isProtectedCommit {
+		return ErrProtectedBranch
+	}
+	isProtectedWrite, err := g.protectedBranchesManager.IsBlocked(ctx, repository, branchID, BranchProtectionBlockedAction_STAGING_WRITE)
+	if err != nil {
+		return err
+	}
+	if isProtectedWrite {
+		return ErrWriteToProtectedBranch
+	}
+
+	options := &SetOptions{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	if repository.ReadOnly && !options.Force {
+		return ErrReadOnlyRepository
+	}
+
+	// TODO(ariels): up to here.  Verify staging is empty!
+	err = g.RefManager.BranchUpdate(ctx, repository, branchID, func(branch *Branch) (*Branch, error) {
+		// Must fetch ref under update!  Consider a hard reset to
+		// branch^ racing against a commit D to branch.  Say branch
+		// looks like this:
+		//
+		//    C --> B --> A
+		//
+		// There are 2 possible results:
+		//
+		// * If commit occurs first then D --> C --> B --> A and
+		//   then the hard reset goes to C --> B --> A.
+		//
+		// * If the hard reset goes first then B --> A and then the
+		//   commit goes to D --> B --> A.
+		//
+		// But a third option is possible if dereference occurs
+		// outside of BranchUpdate.  Hard reset dereferences branch^
+		// to B, then commit goes to D --> C --> B --> A, then hard
+		// reset enters BranchUpdate and we end up with B --> A.
+		// That result is not serialized: it cannot be explained
+		// through atomic operations.
+		//
+		// So we need to dereference here :-(
+		commitRecord, err := g.dereferenceCommit(ctx, repository, ref)
+		if err != nil {
+			return nil, err
+		}
+		if branch.StagingToken != "" || len(branch.SealedTokens) > 0 {
+			return nil, ErrDirtyBranch
+		}
```

Comment:
- Chekcing staging token before dereference will do better
- We have isSealedEmpty

Reply:
Thanks, done!